cloud-sql-connector-r2dbc-postgres unable to connect to DB after SSL certificate rotation

### Bug Description

We are currently using  cloud-sql-connector-r2dbc-postgres v1.15.0 to connect from our java application to Cloud SQL using IAM authentication. The application normally works fine until when we rotate the DB SSL certificate, at which point the application is no longer able to connect to the database. If we rollback the certificate, the connection resumes just fine.

### Example code (or command)

```Java
pom.xml

<?xml version="1.0" encoding="UTF-8"?>

<project xmlns:xsi=http://www.w3.org/2001/XMLSchema-instance xmlns=http://maven.apache.org/POM/4.0.0
         xsi:schemaLocation=http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd>
    <modelVersion>4.0.0</modelVersion>

    <groupId>com.test.cloudsql</groupId>
    <artifactId>test-cloudsql-socketfactory</artifactId>
    <version>${revision}</version>
    <name>${project.artifactId}</name>

    <properties>
        <maven.compiler.release>17</maven.compiler.release>
        <revision>1.0-SNAPSHOT</revision>
    </properties>

    <dependencies>

        <dependency>
            <groupId>io.r2dbc</groupId>
            <artifactId>r2dbc-pool</artifactId>
            <version>1.0.1.RELEASE</version>
        </dependency>

        <dependency>
            <groupId>org.postgresql</groupId>
            <artifactId>r2dbc-postgresql</artifactId>
            <version>1.0.2.RELEASE</version>
        </dependency>

        <dependency>
            <groupId>com.google.cloud.sql</groupId>
            <artifactId>cloud-sql-connector-r2dbc-postgres</artifactId>
            <version>1.15.0</version>
        </dependency>

        <dependency>
            <groupId>org.slf4j</groupId>
            <artifactId>slf4j-simple</artifactId>
            <version>2.0.9</version>
        </dependency>

        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.9.3</version>
            <scope>test</scope>
        </dependency>

        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>5.9.3</version>
            <scope>test</scope>
        </dependency>

    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.11.0</version>
                <configuration>
                    <release>17</release>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>3.1.2</version>
            </plugin>
        </plugins>
    </build>

</project>

Test class:

package com.test.cloudsql;

import static com.google.cloud.sql.core.GcpConnectionFactoryProvider.ENABLE_IAM_AUTH;
import static io.r2dbc.spi.ConnectionFactoryOptions.DATABASE;
import static io.r2dbc.spi.ConnectionFactoryOptions.DRIVER;
import static io.r2dbc.spi.ConnectionFactoryOptions.HOST;
import static io.r2dbc.spi.ConnectionFactoryOptions.PASSWORD;
import static io.r2dbc.spi.ConnectionFactoryOptions.PROTOCOL;
import static io.r2dbc.spi.ConnectionFactoryOptions.USER;

import io.r2dbc.pool.ConnectionPool;
import io.r2dbc.pool.ConnectionPoolConfiguration;
import io.r2dbc.spi.Connection;
import io.r2dbc.spi.ConnectionFactories;
import io.r2dbc.spi.ConnectionFactory;
import io.r2dbc.spi.ConnectionFactoryOptions;
import java.time.Instant;
import java.util.Timer;
import java.util.TimerTask;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Mono;

public class CloudSqlFactoryTest {

    private static final Logger log = LoggerFactory.getLogger(CloudSqlFactoryTest.class);

    private ConnectionPool connectionPool;

    @BeforeEach
    public void setUpPool() {
        // Set up URL parameters
        ConnectionFactoryOptions options = ConnectionFactoryOptions.builder()
                .option(DRIVER, "pool")
                .option(PROTOCOL, "gcp:postgres")
                .option(PASSWORD, "ignored-but-required")
                .option(USER, "<IAM-USER>")
                .option(DATABASE, "<DB>>")
                .option(HOST, "<DB-HOST>>")
                .option(ENABLE_IAM_AUTH, true)
                .build();

        // Initialize connection pool
        ConnectionFactory connectionFactory = ConnectionFactories.get(options);

        ConnectionPoolConfiguration configuration = ConnectionPoolConfiguration
                .builder(connectionFactory)
                .build();

        this.connectionPool = new ConnectionPool(configuration);
    }

    @Test
    public void pooledConnectionTest() throws InterruptedException {
        final var connPool = this.connectionPool;
        TimerTask task = new TimerTask() {
            public void run() {
                log.info("Getting time from db");
                Mono.usingWhen(Mono.fromDirect(connPool.create()), (connection) -> Mono.fromDirect(
                                        connection.createStatement("SELECT NOW() as TS").execute())
                                .flatMapMany(
                                        (result) -> result.map((r, meta) -> r.get("TS", Instant.class)))
                                .next(), Connection::close, (o, throwable) -> o.close(), Connection::close)
                        .doOnNext(ts -> log.info("Got time from db: {}", ts))
                        .doOnError(ex -> log.error("Error getting time from db", ex))
                        .onErrorComplete()
                        .block();
            }
        };
        Timer timer = new Timer("Timer");
        timer.scheduleAtFixedRate(task, 0L, 5000L);

        Thread.sleep(1000L * 900); // sleep for 15 mins
        timer.cancel();
    }
}
```


### Stacktrace

```bash
io.r2dbc.postgresql.PostgresqlConnectionFactory$PostgresConnectionException: [08003] Cannot connect to 192.160.100.102/<unresolved>:3307

15:13:54   at io.r2dbc.postgresql.PostgresqlConnectionFactory.cannotConnect(PostgresqlConnectionFactory.java:188)

15:13:54   at io.r2dbc.postgresql.PostgresqlConnectionFactory.lambda$doCreateConnection$7(PostgresqlConnectionFactory.java:153)

15:13:54   at reactor.core.publisher.Mono.lambda$onErrorMap$28(Mono.java:3783)

15:13:54   at reactor.core.publisher.FluxOnErrorResume$ResumeSubscriber.onError(FluxOnErrorResume.java:94)

15:13:54   at reactor.core.publisher.MonoFlatMap$FlatMapMain.onError(MonoFlatMap.java:180)

15:13:54   at reactor.core.publisher.MonoDelayUntil$DelayUntilCoordinator.onError(MonoDelayUntil.java:200)

15:13:54   at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onError(FluxMapFuseable.java:142)

15:13:54   at reactor.core.publisher.MonoFlatMap$FlatMapMain.secondError(MonoFlatMap.java:241)

15:13:54   at reactor.core.publisher.MonoFlatMap$FlatMapInner.onError(MonoFlatMap.java:315)

15:13:54   at reactor.core.publisher.MonoIgnoreThen$ThenIgnoreMain.onError(MonoIgnoreThen.java:278)

15:13:54   at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:122)

15:13:54   at reactor.core.publisher.MonoCompletionStage$MonoCompletionStageSubscription.apply(MonoCompletionStage.java:71)

15:13:54   at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934)

15:13:54   at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911)

15:13:54   at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510)

15:13:54   at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2162)

15:13:54   at io.r2dbc.postgresql.client.AbstractPostgresSSLHandlerAdapter.completeHandshakeExceptionally(AbstractPostgresSSLHandlerAdapter.java:73)

15:13:54   at io.r2dbc.postgresql.client.AbstractPostgresSSLHandlerAdapter.operationComplete(AbstractPostgresSSLHandlerAdapter.java:55)

15:13:54   at io.netty.util.concurrent.DefaultPromise.notifyListener0(DefaultPromise.java:590)

15:13:54   at io.netty.util.concurrent.DefaultPromise.notifyListeners0(DefaultPromise.java:583)

15:13:54   at io.netty.util.concurrent.DefaultPromise.notifyListenersNow(DefaultPromise.java:559)

15:13:54   at io.netty.util.concurrent.DefaultPromise.notifyListeners(DefaultPromise.java:492)

15:13:54   at io.netty.util.concurrent.DefaultPromise.setValue0(DefaultPromise.java:636)

15:13:54   at io.netty.util.concurrent.DefaultPromise.setFailure0(DefaultPromise.java:629)

15:13:54   at io.netty.util.concurrent.DefaultPromise.tryFailure(DefaultPromise.java:118)

15:13:54   at io.netty.handler.ssl.SslHandler.handleUnwrapThrowable(SslHandler.java:1361)

15:13:54   at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1343)

15:13:54   at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1387)

15:13:54   at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529)

15:13:54   at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468)

15:13:54   at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290)

15:13:54   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444)

15:13:54   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)

15:13:54   at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412)

15:13:54   at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410)

15:13:54   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440)

15:13:54   at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420)

15:13:54   at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919)

15:13:54   at io.netty.channel.epoll.AbstractEpollStreamChannel$EpollStreamUnsafe.epollInReady(AbstractEpollStreamChannel.java:800)

15:13:54   at io.netty.channel.epoll.EpollEventLoop.processReady(EpollEventLoop.java:499)

15:13:54   at io.netty.channel.epoll.EpollEventLoop.run(EpollEventLoop.java:397)

15:13:54   at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997)

15:13:54   at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74)

15:13:54   at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)

15:13:54   at java.base/java.lang.Thread.run(Thread.java:833)

15:13:54  Caused by: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

15:13:54   at java.base/sun.security.ssl.Alert.createSSLException(Alert.java:131)

15:13:54   at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:371)

15:13:54   at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:314)

15:13:54   at java.base/sun.security.ssl.TransportContext.fatal(TransportContext.java:309)

15:13:54   at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:654)

15:13:54   at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(CertificateMessage.java:473)

15:13:54   at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(CertificateMessage.java:369)

15:13:54   at java.base/sun.security.ssl.SSLHandshake.consume(SSLHandshake.java:396)

15:13:54   at java.base/sun.security.ssl.HandshakeContext.dispatch(HandshakeContext.java:480)

15:13:54   at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1277)

15:13:54   at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask$DelegatedAction.run(SSLEngineImpl.java:1264)

15:13:54   at java.base/java.security.AccessController.doPrivileged(AccessController.java:712)

15:13:54   at java.base/sun.security.ssl.SSLEngineImpl$DelegatedTask.run(SSLEngineImpl.java:1209)

15:13:54   at io.netty.handler.ssl.SslHandler.runDelegatedTasks(SslHandler.java:1651)

15:13:54   at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1497)

15:13:54   at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1338)

15:13:54   ... 18 more

15:13:54  Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

15:13:54   at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:439)

15:13:54   at java.base/sun.security.validator.PKIXValidator.engineValidate(PKIXValidator.java:306)

15:13:54   at java.base/sun.security.validator.Validator.validate(Validator.java:264)

15:13:54   at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(X509TrustManagerImpl.java:285)

15:13:54   at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(X509TrustManagerImpl.java:144)

15:13:54   at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(CertificateMessage.java:632)

15:13:54   ... 29 more

15:13:54  Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target

15:13:54   at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(SunCertPathBuilder.java:141)

15:13:54   at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(SunCertPathBuilder.java:126)

15:13:54   at java.base/java.security.cert.CertPathBuilder.build(CertPathBuilder.java:297)

15:13:54   at java.base/sun.security.validator.PKIXValidator.doBuild(PKIXValidator.java:434)

15:13:54   ... 34 more
```


### Steps to reproduce?

1. Create a sample application with the provided pom.xml and a Junit test class
2. Run the test (which will connect to the database every 5 seconds for 15 mins)
3. Manually rotate the DB SSL certificate
4. Observe from error stack trace from the log
5. Rollback DB SSL certificate
6. Test is able to connect to the database 

### Environment

1. OS type and version: Ubuntu
2. Java SDK version: 17
3. Cloud SQL Java Socket Factory version: 1.15.0


### Additional Details

_No response_

--- a/core/src/main/java/com/google/cloud/sql/core/ConnectionInfoRepositoryFactory.java
+++ b/core/src/main/java/com/google/cloud/sql/core/ConnectionInfoRepositoryFactory.java
@@ -20,8 +20,7 @@ import com.google.api.client.http.HttpRequestInitializer;
 import com.google.cloud.sql.ConnectorConfig;
 
 /** Factory interface for creating SQLAdmin clients to interact with Cloud SQL Admin API. */
-public interface ConnectionInfoRepositoryFactory {
+interface ConnectionInfoRepositoryFactory {
 
-  DefaultConnectionInfoRepository create(
-      HttpRequestInitializer credentials, ConnectorConfig config);
+  ConnectionInfoRepository create(HttpRequestInitializer credentials, ConnectorConfig config);
 }

--- a/core/src/main/java/com/google/cloud/sql/core/Connector.java
+++ b/core/src/main/java/com/google/cloud/sql/core/Connector.java
@@ -35,7 +35,7 @@ import org.slf4j.LoggerFactory;
 class Connector {
   private static final Logger logger = LoggerFactory.getLogger(Connector.class);
 
-  private final DefaultConnectionInfoRepository adminApi;
+  private final ConnectionInfoRepository adminApi;
   private final CredentialFactory instanceCredentialFactory;
   private final ListeningScheduledExecutorService executor;
   private final ListenableFuture<KeyPair> localKeyPair;

--- a/core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java
+++ b/core/src/main/java/com/google/cloud/sql/core/InternalConnectorRegistry.java
@@ -191,6 +191,16 @@ public final class InternalConnectorRegistry {
     return getConnector(config).getConnection(config).getConnectionMetadata(connectTimeoutMs);
   }
 
+  /** Internal use only: Force refresh the connection info. */
+  public void forceRefresh(ConnectionConfig config) {
+    if (config.getNamedConnector() != null) {
+      Connector connector = getNamedConnector(config.getNamedConnector());
+      connector.getConnection(config.withConnectorConfig(connector.getConfig())).forceRefresh();
+    } else {
+      getConnector(config).getConnection(config).forceRefresh();
+    }
+  }
+
   private static KeyPair generateRsaKeyPair() {
     KeyPairGenerator generator;
     try {

--- a/core/src/test/java/com/google/cloud/sql/core/DefaultConnectionInfoRepositoryTest.java
+++ b/core/src/test/java/com/google/cloud/sql/core/DefaultConnectionInfoRepositoryTest.java
@@ -53,7 +53,7 @@ public class DefaultConnectionInfoRepositoryTest {
     MockAdminApi mockAdminApi =
         buildMockAdminApi(INSTANCE_CONNECTION_NAME, DATABASE_VERSION, DEFAULT_BASE_URL);
     ConnectorConfig config = new ConnectorConfig.Builder().build();
-    DefaultConnectionInfoRepository repo =
+    ConnectionInfoRepository repo =
         new StubConnectionInfoRepositoryFactory(mockAdminApi.getHttpTransport())
             .create(new StubCredentialFactory().create(), config);
 
@@ -90,7 +90,7 @@ public class DefaultConnectionInfoRepositoryTest {
         INSTANCE_CONNECTION_NAME, Duration.ofHours(1), DEFAULT_BASE_URL);
     ConnectorConfig config = new ConnectorConfig.Builder().build();
 
-    DefaultConnectionInfoRepository repo =
+    ConnectionInfoRepository repo =
         new StubConnectionInfoRepositoryFactory(mockAdminApi.getHttpTransport())
             .create(new StubCredentialFactory().create(), config);
 
@@ -124,7 +124,7 @@ public class DefaultConnectionInfoRepositoryTest {
     MockAdminApi mockAdminApi =
         buildMockAdminApi(INSTANCE_CONNECTION_NAME, "SQLSERVER_2019_STANDARD", DEFAULT_BASE_URL);
     ConnectorConfig config = new ConnectorConfig.Builder().build();
-    DefaultConnectionInfoRepository repo =
+    ConnectionInfoRepository repo =
         new StubConnectionInfoRepositoryFactory(mockAdminApi.getHttpTransport())
             .create(new StubCredentialFactory().create(), config);
 
@@ -151,7 +151,7 @@ public class DefaultConnectionInfoRepositoryTest {
     MockAdminApi mockAdminApi =
         buildMockAdminApi(INSTANCE_CONNECTION_NAME, DATABASE_VERSION, DEFAULT_BASE_URL);
     ConnectorConfig config = new ConnectorConfig.Builder().build();
-    DefaultConnectionInfoRepository repo =
+    ConnectionInfoRepository repo =
         new StubConnectionInfoRepositoryFactory(new BadConnectionFactory())
             .create(new StubCredentialFactory().create(), config);
 
@@ -188,7 +188,7 @@ public class DefaultConnectionInfoRepositoryTest {
             .withAdminRootUrl(adminRootUrl)
             .withAdminServicePath(adminServicePath)
             .build();
-    DefaultConnectionInfoRepository repo =
+    ConnectionInfoRepository repo =
         new StubConnectionInfoRepositoryFactory(mockAdminApi.getHttpTransport())
             .create(new StubCredentialFactory().create(), config);
 

--- a/core/src/test/java/com/google/cloud/sql/core/InternalConnectorRegistryTest.java
+++ b/core/src/test/java/com/google/cloud/sql/core/InternalConnectorRegistryTest.java
@@ -35,6 +35,7 @@ import java.time.Duration;
 import java.time.Instant;
 import java.util.Collections;
 import java.util.Properties;
+import java.util.concurrent.TimeoutException;
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
@@ -445,6 +446,39 @@ public class InternalConnectorRegistryTest extends CloudSqlCoreTestingBase {
     IllegalArgumentException ex =
         assertThrows(IllegalArgumentException.class, () -> registry.connect(nameOnlyConfig));
     assertThat(ex).hasMessageThat().contains("Named connection my-connection does not exist.");
+  }
+
+  @Test
+  public void forceRefreshTest() throws IOException, InterruptedException, TimeoutException {
+    final String namedConnector = "connection-internal";
+    final PauseCondition refresh = new PauseCondition();
+    FakeSslServer sslServer = new FakeSslServer();
+    int port = sslServer.start(PUBLIC_IP);
+    StubConnectionInfoRepository connectionInfoRepository = new StubConnectionInfoRepository();
+    ConnectionInfoRepositoryFactory factory =
+        new StubConnectionInfoRepositoryFactory(connectionInfoRepository);
+    InternalConnectorRegistry registry =
+        new InternalConnectorRegistry(
+            clientKeyPair,
+            factory,
+            stubCredentialFactoryProvider,
+            port,
+            TEST_MAX_REFRESH_MS,
+            defaultExecutor);
+
+    ConnectionConfig config =
+        new ConnectionConfig.Builder()
+            .withCloudSqlInstance("myProject:myRegion:myInstance")
+            .withNamedConnector(namedConnector)
+            .build();
+
+    ConnectorConfig configWithDetails = new ConnectorConfig.Builder().build();
+    registry.register(namedConnector, configWithDetails);
+    assertThat(connectionInfoRepository.getRefreshCount()).isEqualTo(0);
+
+    registry.forceRefresh(config);
+    refresh.waitForCondition(() -> connectionInfoRepository.getRefreshCount() > 0, 1000);
+    assertThat(connectionInfoRepository.getRefreshCount()).isEqualTo(1);
   }
 
   private InternalConnectorRegistry createRegistry(

--- a/core/src/test/java/com/google/cloud/sql/core/StubConnectionInfoRepository.java
+++ b/core/src/test/java/com/google/cloud/sql/core/StubConnectionInfoRepository.java
@@ -1,0 +1,64 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import com.google.cloud.sql.AuthType;
+import com.google.cloud.sql.IpType;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningScheduledExecutorService;
+import java.security.KeyPair;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
+import javax.net.ssl.KeyManagerFactory;
+
+class StubConnectionInfoRepository implements ConnectionInfoRepository {
+  private AtomicInteger refreshCount = new AtomicInteger();
+
+  int getRefreshCount() {
+    return refreshCount.get();
+  }
+
+  private ConnectionInfo newConnectionInfo() {
+    Map<IpType, String> ips = Collections.singletonMap(IpType.PUBLIC, "10.1.1.1");
+    try {
+      return new ConnectionInfo(
+          new InstanceMetadata(ips, null),
+          new SslData(
+              null, KeyManagerFactory.getInstance(KeyManagerFactory.getDefaultAlgorithm()), null),
+          Instant.now().plus(1, ChronoUnit.HOURS));
+    } catch (NoSuchAlgorithmException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public ListenableFuture<ConnectionInfo> getConnectionInfo(
+      CloudSqlInstanceName instanceName,
+      AccessTokenSupplier accessTokenSupplier,
+      AuthType authType,
+      ListeningScheduledExecutorService executor,
+      ListenableFuture<KeyPair> keyPair) {
+    refreshCount.incrementAndGet();
+    ConnectionInfo connectionInfo = newConnectionInfo();
+    return Futures.immediateFuture(connectionInfo);
+  }
+}

--- a/core/src/test/java/com/google/cloud/sql/core/StubConnectionInfoRepositoryFactory.java
+++ b/core/src/test/java/com/google/cloud/sql/core/StubConnectionInfoRepositoryFactory.java
@@ -26,14 +26,23 @@ import com.google.cloud.sql.ConnectorConfig;
 public class StubConnectionInfoRepositoryFactory implements ConnectionInfoRepositoryFactory {
 
   HttpTransport httpTransport;
+  ConnectionInfoRepository connectionInfoRepository;
 
   StubConnectionInfoRepositoryFactory(HttpTransport transport) {
     this.httpTransport = transport;
   }
 
+  StubConnectionInfoRepositoryFactory(ConnectionInfoRepository connectionInfoRepository) {
+    this.connectionInfoRepository = connectionInfoRepository;
+  }
+
   @Override
-  public DefaultConnectionInfoRepository create(
+  public ConnectionInfoRepository create(
       HttpRequestInitializer credentials, ConnectorConfig config) {
+    if (connectionInfoRepository != null) {
+      return connectionInfoRepository;
+    }
+
     SQLAdmin.Builder adminApiBuilder =
         new SQLAdmin.Builder(
                 httpTransport != null ? httpTransport : new MockHttpTransport(),

--- a/r2dbc/core/src/main/java/com/google/cloud/sql/core/CloudSqlConnection.java
+++ b/r2dbc/core/src/main/java/com/google/cloud/sql/core/CloudSqlConnection.java
@@ -1,0 +1,149 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import io.r2dbc.spi.Batch;
+import io.r2dbc.spi.Connection;
+import io.r2dbc.spi.ConnectionMetadata;
+import io.r2dbc.spi.IsolationLevel;
+import io.r2dbc.spi.Statement;
+import io.r2dbc.spi.TransactionDefinition;
+import io.r2dbc.spi.ValidationDepth;
+import java.time.Duration;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+/** {@link CloudSqlConnection} for connecting to Cloud SQL instances via R2DBC protocol. */
+public class CloudSqlConnection implements Connection {
+  private final Connection delegate;
+  private final ConnectionConfig config;
+
+  /** Creates an instance of Connection. */
+  public CloudSqlConnection(ConnectionConfig config, Connection delegate) {
+    this.config = config;
+    this.delegate = delegate;
+  }
+
+  @Override
+  public Publisher<Void> setStatementTimeout(Duration timeout) {
+    return delegate.setStatementTimeout(timeout);
+  }
+
+  @Override
+  public Publisher<Void> beginTransaction() {
+    return delegate.beginTransaction();
+  }
+
+  @Override
+  public Publisher<Void> beginTransaction(TransactionDefinition definition) {
+    return delegate.beginTransaction(definition);
+  }
+
+  @Override
+  public Publisher<Void> close() {
+    return delegate.close();
+  }
+
+  @Override
+  public Publisher<Void> commitTransaction() {
+    return delegate.commitTransaction();
+  }
+
+  @Override
+  public Batch createBatch() {
+    return delegate.createBatch();
+  }
+
+  @Override
+  public Publisher<Void> createSavepoint(String name) {
+    return delegate.createSavepoint(name);
+  }
+
+  @Override
+  public Statement createStatement(String sql) {
+    return delegate.createStatement(sql);
+  }
+
+  @Override
+  public boolean isAutoCommit() {
+    return delegate.isAutoCommit();
+  }
+
+  @Override
+  public ConnectionMetadata getMetadata() {
+    return delegate.getMetadata();
+  }
+
+  @Override
+  public IsolationLevel getTransactionIsolationLevel() {
+    return delegate.getTransactionIsolationLevel();
+  }
+
+  @Override
+  public Publisher<Void> releaseSavepoint(String name) {
+    return delegate.releaseSavepoint(name);
+  }
+
+  @Override
+  public Publisher<Void> rollbackTransaction() {
+    return delegate.rollbackTransaction();
+  }
+
+  @Override
+  public Publisher<Void> rollbackTransactionToSavepoint(String name) {
+    return delegate.rollbackTransactionToSavepoint(name);
+  }
+
+  @Override
+  public Publisher<Void> setAutoCommit(boolean autoCommit) {
+    return delegate.setAutoCommit(autoCommit);
+  }
+
+  @Override
+  public Publisher<Void> setTransactionIsolationLevel(IsolationLevel isolationLevel) {
+    return delegate.setTransactionIsolationLevel(isolationLevel);
+  }
+
+  @Override
+  public Publisher<Boolean> validate(ValidationDepth depth) {
+    Publisher<Boolean> response = delegate.validate(depth);
+    Mono.from(response)
+        .doOnNext(
+            valid -> {
+              if (!valid) {
+                // Force refresh the connection info when it is invalid.
+                InternalConnectorRegistry.getInstance().forceRefresh(config);
+              }
+            })
+        // Execute in a default scheduler to prevent blocking.
+        .subscribeOn(Schedulers.boundedElastic())
+        .share()
+        .block();
+    return response;
+  }
+
+  @Override
+  public String toString() {
+    return "CloudSqlConnection{" + ", delegate=" + delegate + '}';
+  }
+
+  @Override
+  public Publisher<Void> setLockWaitTimeout(Duration timeout) {
+    return delegate.setLockWaitTimeout(timeout);
+  }
+}

--- a/r2dbc/core/src/main/java/com/google/cloud/sql/core/CloudSqlConnectionFactory.java
+++ b/r2dbc/core/src/main/java/com/google/cloud/sql/core/CloudSqlConnectionFactory.java
@@ -26,6 +26,7 @@ import io.r2dbc.spi.ConnectionFactoryOptions;
 import io.r2dbc.spi.ConnectionFactoryProvider;
 import java.util.function.Supplier;
 import org.reactivestreams.Publisher;
+import reactor.core.publisher.Mono;
 import reactor.util.annotation.NonNull;
 
 /** {@link ConnectionFactory} for accessing Cloud SQL instances via R2DBC protocol. */
@@ -54,7 +55,9 @@ public class CloudSqlConnectionFactory implements ConnectionFactory {
             .getConnectionMetadata(config)
             .getPreferredIpAddress();
     builder.option(HOST, hostIp).option(PORT, SERVER_PROXY_PORT);
-    return supplier.get().create(builder.build()).create();
+
+    return Mono.from(supplier.get().create(builder.build()).create())
+        .map(c -> new CloudSqlConnection(config, c));
   }
 
   @Override

